### ISSUE TITLE
dhcpcd is installed even on systems using dhclient (Debian)

### ISSUE THREAD

**Issue by hkbakke** [pi-hole/pi-hole#2989]
Installed Pi-hole on Debian Buster. After a while my internet connectivity broke. The reason being that dhcpcd is installed, and it had actually gotten a second DHCP-lease on my WAN link, while the default Debian dhclient had gotten the first. dhcpcd makes sense on raspbian, but it should perhaps not be installed on full blown Debian systems already having dhclient (the default).

I did a couple of workarounds:

Add `denyinterfaces *` to /etc/dhcpcd.conf.

Or the more useful:

    systemctl stop dhcpcd.service
    systemctl disable dhcpcd.service

Not sure if the last one will survive every upgrade in the future though. Did the same for resolvconf which broke DNS too.

So in short, both DHCP and DNS breaks either subtly or completely when Pi-hole is installed on full blown Debian by assuming what works on raspbian makes sense in ordinary Debian server installs.

**Comment by dschaper** [comment]
`dhcpcd` is intentionally installed as that is how Raspbian sets a static IP address, as you have noted. 

Another good topic for discussion would be the need for hand holding users to set static IP addresses versus requiring the users to handle IP addressing themselves. Since Raspbian uses `dhcpcd` as the addressing scheme, do we even need to have that as a dependency to install and add some more intelligence to the installer so that it will only offer to set the static IP if it detects `dhcpcd` pre-installed. 

There are a large number of users that may not know how to set the IP address or assume that since the DHCP server on their routers handles the task then they don't need to worry until the lease changes and the IP changes thus taking down DNS for the network.

**Comment by hkbakke** [comment]
I am of the same view that it is not expected that a service does this for you. It is perhaps an option to do this automatically on raspbian, which covers the "standard" Pi-hole install, but leave it up to the user on other platforms where you know a lot less about your surroundings.

**Comment by dschaper** [comment]
There are so many avenues to cover (`dhclient`, `networkmanager`, `dhcpcd`, `netplan`) that I think it may just be too much to try to work in to an intelligent installer. As for your experience, have you considered `docker` as an option? Solves a lot of the common issues by going that route.

**Comment by hkbakke** [comment]
I do run many dockers, but in this case it was not a good match as I decided to run Pi-hole on my gateway which has a complex firewall rule set in nftables (the default in Debian Buster) which is not really compatible with docker at the moment. Also I use the DHCP-server and am not comfortable with it claiming port 80, so I am more likely to move to a VM instead.
I pretty much expected Pi-hole to be dnsmasq tweaked with a admin GUI, and not a complete network configurator for your system though, so if the resolvconf, dhcpcd thing continues to make trouble for me I will probably put it in a dedicated VM. I am a system administrator by trade, so I am able to work around the current issues my self. This issue is more for everyone else, as the current "configure the system in addition to dnsmasq" approach is very likely to behave somewhat erratically on all but raspbian default setups, but people might not understand why. 

In short I think it is easier to figure out/learn by some googling how to set DNS and static IP for your chosen OS platform as a novice, than figure out what does not work when it breaks.

**Comment by DL6ER** [comment]
> In short I think it is easier to figure out/learn by some googling how to set DNS and static IP for your chosen OS platform as a novice, than figure out what does not work when it breaks (or rather works only some of the time)

Well, I disagree. Strongly. For two reasons.
First: Or user base contains *a lot* Linux first timers. Many told us that they use Pi-hole as their first computer project beyond Chrome and Office *at all*. We are regularly helping users that do not even know how to copy/paste things from our support forum into their terminal or vice versa. We want to provide a good and easy Installation process for them and I think we're doing well.

Second: It may not be working for you but you're also trying to install Pi-hole on your gateway which was never be planned to work at all. The installer had been built for Raspberry Pi installations and been tweaked over the years to work also for many other cases (Plain Debian and Fedora, for instance). So please don't generalize your impression. What the installer does works for the vast majority of our users out of the box.

@dschaper is right, we've discussed this multiple times in the past. My impression has always been that we should make it easy for first time SBC / Linux users not even knowing that there are operating systems out there that could be used without a screen. Another option dialog in the installer is fine, but the Enter mashing many people do will only change one thing: We could tell them (in the end): but the installer asked you what to do. Would this really help them any further? I doubt.

**Comment by hkbakke** [comment]
Actually, the real issue is that there is no opt out during install. That single option would solve this for both camps.

A single question like e.g "Do you want to configure IP and DNS on your system", with the default being Yes for the enter mashers (the typical user), and No for the people that needing to set this themself.

Also, regarding Debian specifically, I consider it pure luck if the current approach with dhcpcd and resolvconf works without issues with the default network/dns stuff on the system. The only way I can see that working as intended is on a single NIC server with no existing network or DNS config, and it is still just a mess waiting to surface in the future. I think that it would be better not to configure Debian systems at all with the current approach....or give us an opt out possibility.

**Comment by DL6ER** [comment]
> That single option would solve this for both camps.

Yes. This is what I meant. We'll work on this if there are no objections from the rest of the team.

**Comment by hkbakke** [comment]
Until (If) that happens I just created a fork and removed the resolvconf and dhcpcd packages from the autoinstaller script (and removed the resolvconf dns server update from the init.d script). It allows me to upgrade the install without it reinstalling thos packages.

https://github.com/hkbakke/pi-hole/tree/no-system-net-and-dns-config

I have bought a RPi4 for the purpose of Pi-hole though. I was taken by surprise that it configured the whole system, so I'll rather have it isolated to a place where it can't do any damage. I'll probably run it in a docker too, as the base system will be managed by a configuration management system.

On a side note:
As I was looking through your installer script I noticed that a lot of what you are doing by bash scripting with the complexity that follows, is actually stuff that configuration management systems already does for you, but in a much clearer way, and mostly platform independent. I know several other projects that use masterless configuration management systems to handle their complex install. Gitlab (Chef) and Qubes (Saltstack) coming to mind. It seems like the Pi-hole project is really having more the "appliance" mindset, and not "just another service on the box", so it might be something to consider in the future instead of adding massive amounts of shell scripting cruft that needs continuous follow up for different platforms etc.

**Comment by dschaper** [comment]
I don't think learning a CMS (puppet,chef,ansible,saltstack) is going to be a good prerequisite for users to learn before installing Pi-hole.

**Comment by hkbakke** [comment]
Perhaps it wasn't clear, but it is not for the users, it's for the installer script. Instead of writing a lot of bash code teaching the installer how to set IP-addresses, start services etc, you could instead just create e.g a state that says: install dnsmasq and lighthttpd, add this templated config, ensure service is started in a high level language like yaml, and then something like salt takes care of how to do it, regardless of platform.

**Comment by dschaper** [comment]
I understood, my point is that then the user would need to install and manage a CMS. Set up a puppet master, or create a Tower for Ansible to manage updates and the like. 

Pi-hole is an appliance, if you want to do things in a managed setup or on a server then use `docker`.

**Comment by hkbakke** [comment]
No, i still don't think you understand. It is the installer scripts that installs a masterless CMS where the entire configuration is kept within the installer. No external daemons, CMS setup for the user etc. I am pretty sure it would drastically simplify the installer process for you and simplify the maintenance. Puppet is obviously not a good choice, but something like Salt is.

**Comment by dschaper** [comment]
Yeah, that's not going to work well on devices or installs where there is a bare minimum OS and 512M of RAM. We go from managing things with bash to managing the management system with bash and add a huge layer of bloat and complexity. I can see people asking "This is a DNS, why do you need to install Salt!?" Users are asking us to remove things like `dialog` from the package, I really don't think they would be good with installing a CMS.

**Comment by hkbakke** [comment]
It was just a suggestion anyway. It makes sense if you do the appliance approach where you want to manage several services and parts of the system. It's not feasible to do it that way unless the developers is comfortable with the CMS anyway. The system minimum requirements is not really an issue if using the right tool for the job (salt-minion). Raspberry Pi has plenty of resources for something like salt-minion, but it's of course always nice to keep the number of dependencies down. Which is actually another argument in favour of not touching the network and DNS settings, as the rather unique "also touch the system" approach in the installer requires extra dependencies.

**Comment by dschaper** [comment]
We have to keep in mind users that have never seen a linux system before in their lives. We are a unique project and fit a unique segment. If you are new to things then you need the extra help. If you are a seasoned admin then use the `docker` repo.

**Comment by Jorgen-P** [comment]
> We have to keep in mind users that have never seen a linux system before in their lives. We are a unique project and fit a unique segment. If you are new to things then you need the extra help. If you are a seasoned admin then use the `docker` repo.

Is this really the policy? That if I don't want my network config demolished I have to use Docker? There seems to be no option in the install script to simply leave the network config alone and just install Pihole.

**Comment by PromoFaux** [comment]
Fret not, it's being worked on:

https://github.com/pi-hole/pi-hole/pull/3715

**Comment by Jorgen-P** [comment]
Ok, great. Thanks for fixing this.

**Comment by fclaerho** [comment]
I agree with the original report; I just recently installed pihole on my home server (Ubuntu) given it's a popular ad-blocker, without docker, and was pretty surprised by the many side effects, I was just expecting it to be a classic service sitting on port 53 plus a web app, but the installation process is quite involved and I had to tweak things manually and read a bunch of code to understand what was happening — the installation process tries to kick out systemd-resolved but does not do it entirely (issue already reported with the /etc/resolv.conf symlink) then come dhcpcd which is overwritting that same file. 

I understand that the nominal use case here is pihole alone on a raspberry pi, but the project is also advertised to work on many platforms. I'd largely prefer the installer to just install its bits and not tinker with the system configuration when it's not in this nominal use case :[

For instance, given my router is already configured to allocate a static IP via DHCP to my home server, I do not need this additional dhcpcd configuration.

That being said, once up and running, pihole is sweet, thanks for the work! I just wish the installer would be largely simplified tho and not try to cover an impossible number of use cases.

**Comment by PromoFaux** [comment]
We hear you. I need to try and re-motivate myself to getting back to the above-mentioned PR, but we definitely agree that it is currently not optimal

**Comment by yubiuser** [comment]
This PR will stop installation of `dhcpd5`: https://github.com/pi-hole/pi-hole/pull/4260

**Comment by yubiuser** [comment]
Fixed with current release
https://discourse.pi-hole.net/t/pi-hole-ftl-v5-9-web-v5-6-and-core-v5-4-released/49544
